### PR TITLE
Add some concept portability macros to use concepts when they are available

### PR DESCRIPTION
**Pull request by ericniebler**
Also, replace many `enable_if`'s with `requires` clauses.

### CODE REVIEW

**Comment by ericniebler** [comment]
I tried to leave alone the places where `enable_if` was being used to select between overloads. Only places that were using `enable_if` purely to constrain functions have been switched to `requires` clauses via the `UNIFEX_TEMPLATE` syntax.

This will also give us a way to turn off checking. I can add a mode where `UNIFEX_TEMPLATE` strips out the `requires` clauses entirely, turning shallow errors into deep ones.

File: include/unifex/sequence.hpp [hunk 1]
```diff
@@ -185,36 +176,25 @@ namespace unifex
         }
       }
 
-      template <
-          typename Error,
-          std::enable_if_t<
-              is_callable_v<decltype(unifex::set_error), Receiver, Error>,
-              int> = 0>
+      UNIFEX_TEMPLATE(typename Error)
+          (requires is_callable_v<decltype(unifex::set_error), Receiver, Error>)
```
Comment (kirkshoop) [inline include/unifex/sequence.hpp]: use callable concept instead of trait?

File: include/unifex/sequence.hpp [hunk 2]
```diff
@@ -185,36 +176,25 @@ namespace unifex
...
             static_cast<Error&&>(error));
       }
 
-      template <
-          typename... Args,
-          std::enable_if_t<
-              is_callable_v<decltype(unifex::set_done), Receiver, Args...>,
-              int> = 0>
+      UNIFEX_TEMPLATE(typename... Args)
+          (requires is_callable_v<decltype(unifex::set_done), Receiver, Args...>)
```
Comment (kirkshoop) [inline include/unifex/sequence.hpp]: use callable concept instead of trait?

File: include/unifex/sequence.hpp [hunk 3]
```diff
@@ -333,13 +313,9 @@ namespace unifex
...
-      template <
-          typename Predecessor2,
-          typename Successor2,
-          std::enable_if_t<
-              std::is_constructible_v<Predecessor, Predecessor2> &&
-                  std::is_constructible_v<Successor, Successor2>,
-              int> = 0>
+      UNIFEX_TEMPLATE(typename Predecessor2, typename Successor2)
+          (requires std::is_constructible_v<Predecessor, Predecessor2> &&
+              std::is_constructible_v<Successor, Successor2>)
```
Comment (kirkshoop) [inline include/unifex/sequence.hpp:318]: constructible_from concept?

File: include/unifex/span.hpp [hunk 4]
```diff
@@ -160,13 +157,9 @@ struct span<T, dynamic_extent> {
...
-  template <
-      typename U,
-      std::size_t OtherExtent,
-      std::enable_if_t<
-          std::is_same_v<U, T> ||
-              (!std::is_const_v<U> && std::is_same_v<const U, T>),
-          int> = 0>
+  UNIFEX_TEMPLATE(typename U, std::size_t OtherExtent)
+      (requires std::is_same_v<U, T> ||
+          ((!std::is_const_v<U>) && std::is_same_v<const U, T>))
```
Comment (kirkshoop) [inline include/unifex/span.hpp:162]: same_as concept?

File: include/unifex/task.hpp [hunk 5]
```diff
@@ -60,9 +61,8 @@ struct task {
       state_ = state::exception;
     }
 
-    template <
-        typename Value,
-        std::enable_if_t<std::is_convertible_v<Value, T>, int> = 0>
+    UNIFEX_TEMPLATE(typename Value)
+        (requires std::is_convertible_v<Value, T>)
```
Comment (kirkshoop) [inline include/unifex/task.hpp:65]: convertible_to concept?

File: include/unifex/transform_done.hpp [hunk 6]
```diff
@@ -347,18 +339,15 @@ inline constexpr struct transform_done_cpo {
...
-  template<
-    typename Source,
-    typename Done,
-    std::enable_if_t<
-        !is_tag_invocable_v<transform_done_cpo, Source, Done> &&
-        std::is_constructible_v<std::remove_cvref_t<Source>, Source> &&
-        std::is_constructible_v<std::remove_cvref_t<Done>, Done> &&
-        is_callable_v<std::remove_cvref_t<Done>>, int> = 0>
+  UNIFEX_TEMPLATE(typename Source, typename Done)
+    (requires (!is_tag_invocable_v<transform_done_cpo, Source, Done>) &&
```
Comment (kirkshoop) [inline include/unifex/transform_done.hpp:343]: tag_invocable concept?

File: include/unifex/typed_via.hpp [hunk 7]
```diff
@@ -39,12 +34,8 @@ namespace _typed_via {
...
     }
 
-    template <
-        typename Source,
-        typename Scheduler,
-        std::enable_if_t<
-            !is_tag_invocable_v<_fn, Source, Scheduler>,
-            int> = 0>
+    UNIFEX_TEMPLATE(typename Source, typename Scheduler)
+      (requires (!is_tag_invocable_v<_fn, Source, Scheduler>))
```
Comment (kirkshoop) [inline include/unifex/typed_via.hpp:38]: tag_invocable concept?

File: include/unifex/dematerialize.hpp [hunk 8]
```diff
@@ -37,54 +38,42 @@ namespace _demat {
   template <typename Receiver>
   class _receiver<Receiver>::type {
    public:
-    template <
-      typename Receiver2,
-      std::enable_if_t<std::is_constructible_v<Receiver, Receiver2>, int> = 0>
+    UNIFEX_TEMPLATE(typename Receiver2)
+      (requires std::is_constructible_v<Receiver, Receiver2>)
```
Comment (kirkshoop) [inline include/unifex/dematerialize.hpp]: constructible_from concept?

File: include/unifex/dematerialize.hpp [hunk 9]
```diff
@@ -37,54 +38,42 @@ namespace _demat {
...
 
-    template <
-        typename CPO,
-        typename... Values,
-        std::enable_if_t<
-            is_receiver_cpo_v<CPO> && is_callable_v<CPO, Receiver, Values...>,
-            int> = 0>
+    UNIFEX_TEMPLATE(typename CPO, typename... Values)
+        (requires is_receiver_cpo_v<CPO> &&
+            is_callable_v<CPO, Receiver, Values...>)
```
Comment (kirkshoop) [inline include/unifex/dematerialize.hpp]: use callable concept instead of trait?

File: include/unifex/detail/concept_macros.hpp [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+  /**/
+#define UNIFEX_TEMPLATE_DEF_SFINAE(...) \
+  template<__VA_ARGS__ UNIFEX_TEMPLATE_DEF_SFINAE_AUX_ \
+  /**/
+#define UNIFEX_TEMPLATE_DEF_SFINAE_AUX_(...) , \
+  typename UNIFEX_true_, \
+  std::enable_if_t< \
+    UNIFEX_FORCE_TO_BOOL( \
+      UNIFEX_PP_CAT(UNIFEX_TEMPLATE_SFINAE_AUX_3_, __VA_ARGS__) && UNIFEX_true_{}), \
+    int>> \
```
Comment (kirkshoop) [inline include/unifex/detail/concept_macros.hpp:386]: should the macros use the `int(*)[__LINE__]` pattern to disambiguate overloads?

File: include/unifex/detail/concept_macros.hpp [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+
+#define UNIFEX_TEMPLATE_SFINAE(...) \
+  template<__VA_ARGS__ UNIFEX_TEMPLATE_SFINAE_AUX_ \
+  /**/
+#define UNIFEX_TEMPLATE_SFINAE_AUX_(...) , \
+  typename UNIFEX_true_ = std::true_type, \
+  std::enable_if_t< \
+    UNIFEX_FORCE_TO_BOOL( \
+      UNIFEX_PP_CAT(UNIFEX_TEMPLATE_SFINAE_AUX_3_, __VA_ARGS__) && UNIFEX_true_{}), \
+    int> = 0> \
```
Comment (kirkshoop) [inline include/unifex/detail/concept_macros.hpp:376]: should the macros use the int(*)[__LINE__] pattern to disambiguate overloads?

File: include/unifex/retry_when.hpp [hunk 12]
```diff
@@ -355,19 +351,17 @@ class _sender<Source, Func>::type {
...
+    (requires constructible_from<Source, const Source&> &&
+        constructible_from<Func, const Func&> &&
+        constructible_from<std::remove_cvref_t<Receiver>, Receiver> &&
+        sender_to<
+            Source&,
+            source_receiver<Source, Func, std::remove_cvref_t<Receiver>>>)
+  operation<Source, Func, Receiver> connect(Receiver&& r) const &
+      noexcept(std::is_nothrow_constructible_v<
+          operation<Source, Func, Receiver>, const Source&, const Func&, Receiver> &&
+          is_nothrow_connectable_v<Source&, source_receiver<Source, Func, std::remove_cvref_t<Receiver>>>) {
```
Comment (kirkshoop) [inline include/unifex/retry_when.hpp:364]: I would like a nothrow trait that is directly relatable to sender_to in the requires expression..

Comment (ericniebler) [reply include/unifex/retry_when.hpp:363]: It's not clear to me that that's an improvement. `sender_to` is a cluster of requirements. If we had an `is_nothrow_sender_to`, which required expression should be noexcept? All of them, it just `connect`? Why?

File: include/unifex/any_unique.hpp [hunk 13]
```diff
@@ -319,13 +319,10 @@ class _make<CPOs...>::type
...
-      typename Concrete,
-      typename Allocator,
-      std::enable_if_t<
-          !(std::is_same_v<std::allocator_arg_t, std::decay_t<Concrete>> ||
-            instance_of_v<std::in_place_type_t, std::decay_t<Concrete>>),
-          int> = 0>
+  UNIFEX_TEMPLATE(typename Concrete, typename Allocator)
+      (requires
+          (!(same_as<std::allocator_arg_t, std::decay_t<Concrete>> ||
+             instance_of_v<std::in_place_type_t, std::decay_t<Concrete>>)))
```
[the remark this answers is not in the document]
Comment (ericniebler) [reply include/unifex/any_unique.hpp:325]: `instance_of` feels more like a trait than a concept.

File: include/unifex/detail/concept_macros.hpp (hunk 10 above)
Comment (ericniebler) [reply include/unifex/detail/concept_macros.hpp:386]: It hasn't been necessary, and I'm not sure why. I'll leave it off until we need it.

**Pull request closed without merging**